mono/utils/mono-tls.h

```diff
@@ -55,7 +55,15 @@ g_static_assert (TLS_KEY_DOMAIN == 0);
 #define mono_native_tls_alloc(key,destructor) ((*(key) = TlsAlloc ()) != TLS_OUT_OF_INDEXES && destructor == NULL)
 #define mono_native_tls_free TlsFree
 #define mono_native_tls_set_value TlsSetValue
-#define mono_native_tls_get_value TlsGetValue
+
+static inline LPVOID mono_native_tls_get_value (MonoNativeTlsKey key)
+{
+	DWORD last_error = GetLastError ();
+	LPVOID result = TlsGetValue (key);
+	g_assert (GetLastError () == 0);
+	SetLastError (last_error);
+	return result;
```

Comment:
Would be nice to skip Set if Gets return same value, i.e. 0.

Reply:
And both variables could be const.

Reply:
If you are going to call GetLastError, then please compare to the original, and only call SetLastError if they vary. For example if it was already zero.
Or maybe remove the second GetLastError call?

Also, maybe, maybe, use a fs/gs write, instead of a function all.
I'm not sure the offset is public though. Need to check.
That could, could make this slowdown a total of two instructions -- two movs, no calls. But I need to check..

Reply:
Is this good enough? https://en.wikipedia.org/wiki/Win32_Thread_Information_Block

Using that, you can change GetLastError and SetLastError to one instruction each, mov to or from fs:34 or gs:68.

Reply:
@jaykrell I rewrote it using the macros used elsewhere. They do the appropriate checks. If we want to access TEB directly on some platforms it could be done in separate PR in the macros. I believe some versions of MSVC headers used to implement `GetLastError` as inline function. The MinGW header don't do that AFAIK.